### ISSUE TITLE
CDN configuration for RTAMO

### ISSUE THREAD

**Issue by willdurand** [mozilla-services/stubattribution#108]
When a new user follows the RTAMO flow (downloading a FF/Windows installer from AMO), the stub service builds a modified installer that gets uploaded to a CDN. From there, the stub service returns the (CDN) URL to the installer, which is then used to download the file.

It'd be interesting to have answers to these questions:

1. how are the folders/URLs on the CDN generated?
2. how long are the files kept on the CDN?
3. can we introduce a `referer` check on the CDN side for RTAMO?

**Comment by willdurand** [comment]
I think we would need some changes in this project to either add a query param to the CDN URLs OR use a different bucket for RTAMO-related installers (e.g., `rtamo-${product}`). After that we could add a `referer` check to the CDN itself.

**Comment by willdurand** [comment]
@moz-astults would you have some time to look into this please?

**Comment by willdurand** [comment]
> I think we would need some changes in this project to either add a query param to the CDN URLs OR use a different bucket for RTAMO-related installers (e.g., `rtamo-${product}`). After that we could add a `referer` check to the CDN itself.

I submitted a patch for the first part of this idea so that we can hopefully unblock ERTAMO. We may want to tweak a few things like using a "rtamo" directory instead of a prefix but that should be trivial. Once we have something like that, which should not have any impact AFAICT, we could have a simple referer check on the CDN itself (in the nginx config).

**Comment by moz-astults** [comment]
> I suppose introducing a new directory in the storage key could potentially allow us to set a different retention policy (e.g., only keep the files in the CDN for an hour or so).

Unfortunately it seems that GCS lifecycle rules apply bucket-wide and don't have the granularity at the folder/prefix level afaict https://cloud.google.com/storage/docs/lifecycle

I'm probably lacking some context, but I'm not sure how far this needs to be taken because even if we implement this referer header check on stubattribution and the nginx "CDN", the GCS bucket is still publicly exposed right? The nginx CDN is just proxying the request to the bucket, but someone could link to the bucket directly if they knew the URL. I'm fine approving the patch you sent (thank you!) as long as it's understood that this only does so much from a protection point of view.

**Comment by willdurand** [comment]
> Unfortunately it seems that GCS lifecycle rules apply bucket-wide and don't have the granularity at the folder/prefix level afaict [cloud.google.com/storage/docs/lifecycle](https://cloud.google.com/storage/docs/lifecycle)

Ok, that's good to know.

> I'm probably lacking some context, but I'm not sure how far this needs to be taken because even if we implement this referer header check on stubattribution and the nginx "CDN", the GCS bucket is still publicly exposed right? The nginx CDN is just proxying the request to the bucket, but someone could link to the bucket directly if they knew the URL. I'm fine approving the patch you sent (thank you!) as long as it's understood that this only does so much from a protection point of view.

I think that's perfectly fine.

If I get it correctly, we should receive a URL with the `cdn.stubdownloader.services.mozilla.com` domain from the stub service, and this CDN url should serve the modified installer from some GCS bucket. Introducing a referer check for this URL is what we need.

If someone gets the URL to the bucket itself, I'd expect this URL to not have a `mozilla.com` (or `mozilla.org`) domain. In this case, it wouldn't be different than downloading the installer once and hosting it on a different domain. The referer checks we have put between AMO and the "CDN stub downloader" are meant to avoid hotlinking an installer using a mozilla domain.

cc @diox

**Comment by willdurand** [comment]
> @willdurand your PR has been merged. For the nginx (cdn) config, were you thinking something like this?
> 
> ```
>     location ~ rtamo- {
>         if ($http_referer !~ ^https://www.mozilla.org/) {
>           return 403;
>         }
>         proxy_pass $gcs_url/{{ .Values.gcsBucket }}$request_uri;
>     } 
> ```
> 
> cc @jbuck to check my logic here. Testing a config like this locally seems to work as expected.

Yeah, something along these lines but nginx config isn't my _forte_

**Comment by willdurand** [comment]
> @willdurand your PR has been merged.

I verified the patch on stage. It does not break things so that looks good to me :)

**Comment by willdurand** [comment]
I also verified the nginx config change on -stage. Looks good.

I'll file a PI Request to have QA verify prod.

**Comment by willdurand** [comment]
The change for this issue (both the stubattribution and nginx changes) have been deployed to prod today and everything looks good. QA will do a more in-depth verification tomorrow.

**Comment by acornestean** [comment]
Based on the testing instructions described in https://gist.github.com/diox/97b4e81a2cd3d66d395fcc4751626061, the 3rd link from the gist should download Firefox but not trigger RTAMO, however, RTAMO is triggered upon browser launch.

As discussed with @willdurand, this is in order for RTAMO, but is a blocking issue for ERTAMO.

Other than this everything works as expected. A short summary of the verification can be seen below:

On Windows I tested 3 browsers (Edge, Chrome and Opera):
1. when starting from a recommended add-on page on AMO, Firefox will get downloaded and installed automaticaly. Upon launch RTAMO is triggered and the served add-on is shown on the about:welcome page and installable, as expected.
2. when starting from a non-recommended add-on page on AMO, Firefox will get downloaded and installed automaticaly. Upon launch RTAMO is NOT triggered, as expected.
3. when using the links from the gist:
- the first link in the gist links to the uBlock AMO page and starting the flow from there will trigger RTAMO as expected.
- the 2nd, 4th and 5th links proceed to Firefox being downloaded but will not trigger RTAMO, as expected per the mentions in the gist.
- the 3rd link however, proceeds to Firefox being downloaded and will trigger RTAMO on browser launch (as mentioned above)

For Linux I checked 2 browsers (Chrome and Opera) and on MacOS I also checked 2 browsers (Chrome and Safari):
1. when starting from a recommended add-on page on AMO, Firefox will get downloaded. After being installed and launched RTAMO is NOT triggered, as expected for Linux and MacOS.
2. when starting from a non-recommended add-on page on AMO, Firefox will get downloaded. After being installed and launched RTAMO is NOT triggered, as expected for Linux and MacOS.
